serverless/virus-scanner/src/index.ts

```diff
@@ -0,0 +1,109 @@
+import { APIGatewayProxyResult } from 'aws-lambda'
+import crypto from 'crypto'
+import { StatusCodes } from 'http-status-codes'
+
+import { scanFileStream } from './clamscan.service'
+import { config } from './config'
+import { getLambdaLogger } from './logger'
+import { S3Service } from './s3.service'
+import { isBodyWithKey } from './types'
+
+export const handler = async (
+  event: unknown,
+): Promise<APIGatewayProxyResult> => {
+  const logger = getLambdaLogger('virus-scanner')
+
+  logger.info('handler triggered')
+
+  if (!isBodyWithKey(event)) {
+    logger.warn({
+      message: 'Missing key in body',
+      event,
+    })
+    return {
+      statusCode: StatusCodes.BAD_REQUEST,
+      body: JSON.stringify({
+        message: 'Missing key in body',
+      }),
+    }
+  }
+
+  const quarantineFileKey = event.key
+
+  const cleanBucket = config.virusScannerCleanS3Bucket
+  const quarantineBucket = config.virusScannerQuarantineS3Bucket
+
+  // Retrieve from S3
+  const s3Client = new S3Service(config.isTestOrDev, logger)
+
+  let s3ReadableStream
+  try {
+    s3ReadableStream = await s3Client.getS3FileStream({
+      bucketName: quarantineBucket,
+      objectKey: quarantineFileKey,
+    })
+  } catch (error) {
+    logger.error({
+      message: 'File not found',
+      error,
+      quarantineFileKey,
+    })
+    return {
+      statusCode: StatusCodes.NOT_FOUND,
+      body: JSON.stringify({
+        message: 'File not found',
+        fileKey: quarantineFileKey,
+      }),
+    }
+  }
+
+  // Scan file
+
+  const scanResult = await scanFileStream(s3ReadableStream)
+  const { isMalicious } = scanResult
+
+  // Move to clean bucket if clean
+  if (isMalicious) {
+    const { virusMetadata } = scanResult
+
+    logger.error({
+      message: 'malicious file detected!!',
+      virusMetadata,
+      quarantineFileKey,
+    })
+    return {
+      statusCode: StatusCodes.BAD_REQUEST,
+      body: JSON.stringify({
+        message: 'Malicious file detected',
+        fileKey: quarantineFileKey,
+      }),
+    }
+  } else {
+    const cleanFileKey = crypto.randomUUID()
+    logger.info({
+      message: 'clean file detected',
+      cleanFileKey,
+    })
+
+    const { cleanFile } = scanResult
+
+    await s3Client.putS3File({
```

Comment:
Same comment here about whether this can be a copy operation instead. It could even be a move operation.

Reply:
[7dea90f](https://github.com/opengovsg/FormSG/pull/6642/commits/7dea90f284efeac09a502e14f65ff1ed4dc17532)